feat: allow setting custom parents via metadata (Alpha) #5082

jake-ciolek opened the pull request:
I took a stab at #5082 and took a slightly different approach. Instead of users specifying the parent/child relationship in config, we could allow setting it in the metadata.

This change allows users to define a parent/child relationship for resources where this cannot be inferred from the ownerRef values. The child resource might be a Cluster-scoped CR and the parent might be a namespaced CRD.

The API is straightforward and boils down to setting annotations that define the parent, i.e:

```
argocd.argoproj.io/parentName: custom-api.example.com/v1alpha1
argocd.argoproj.io/parentKind: Database
argocd.argoproj.io/parentApiVersion: my-claimed-database
argocd.argoproj.io/parentNamespace: parent-namespace
```

If a custom parent is defined, we take it over the ownerRef values.

This allows us to display a more ordered resource tree. Some concrete examples include Istio Operator and Crossplane resources (relationship between Claims and Composite Resources).

In the screenshots, you can see that it brings order to the resource tree of Istio Operator and draws the relationship between Crossplane Claim and Crossplane XR.

This is an early implementation and I would appreciate feedback on this approach and pointers on how to make it better. Thanks!

Updates #5082

<img width="1492" alt="Screenshot 2024-11-22 at 14 34 18" src="https://github.com/user-attachments/assets/9f85a5d5-e1e8-44da-b170-e955311dc399">
<img width="1492" alt="Screenshot 2024-11-22 at 14 34 31" src="https://github.com/user-attachments/assets/d8b3a90b-d7d9-4883-9caa-6974fdcfc2d5">

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [x] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

<!-- Please see [Contribution FAQs](https://argo-cd.readthedocs.io/en/latest/developer-guide/faq/) if you have questions about your pull-request. -->

jake-ciolek committed:
feat: allow setting custom parents via metadata

This change allows users to define a parent/child
relationship for resources where this cannot be inferred
from the ownerRef values. The child resource might be a
Cluster-scoped CR and the parent might be a namespaced CRD.

The API is straightforward and boils down to setting
annotations that define the parent, i.e:

argocd.argoproj.io/ownerApiVersion: custom-api.example.com/v1alpha1
argocd.argoproj.io/ownerKind: Database
argocd.argoproj.io/ownerName: my-claimed-database
argocd.argoproj.io/ownerNamespace: parent-namespace

If a custom parent is defined, we take it over the ownerRef values.

This allows us to display a more ordered resource tree.
Some concrete examples include Istio Operator and Crossplane
resources (relationship between Claims and Composite Resources).

Updates #5082

Signed-off-by: Jakub Ciolek <jakub@ciolek.dev>

andrii-korotkov-verkada commented:
Do you have a gitops-engine PR to update the tree construction logic?

reggie-k reviewed (commented):
Thanks for the PR!
Maybe the annotations should be called "argocd.argoproj.io/parentName", "parentKind", etc, since they are referred to as parents in other parts of the code?
And could you add docs for this in the `docs/user-guide/annotations-and-labels.md`?

jake-ciolek committed:
feat: allow setting custom parents via metadata

- Change annotation names from 'owner*' to 'parent*'.
- Add documentation for the new annotations.
- Fix linter issues.

Signed-off-by: Jakub Ciolek <jakub@ciolek.dev>

jake-ciolek commented:
Thanks @reggie-k . I changed the annotations to parent* and updated the docs.

Aym3nTN commented:
@jake-ciolek @andrii-korotkov-verkada is there any intention to move this forward?

nitishfy reviewed (changes requested):
Thank you for raising the PR. I'd propose few things here:

1. Please resolve the merge conflicts.
2. Regarding the gitops engine comment, the gitops engine code has been migrated to Argo CD code already now. I think resolving the conflicts would give you enough hints to know where to perform additional changes.

Please tag me once you're done resolving the comments. I'd be happy to review this PR and move it forward. Thanks!
